src/ripple/app/misc/impl/AMMUtils.cpp

```diff
@@ -348,4 +348,56 @@ initializeFeeAuctionVote(
         auctionSlot.makeFieldAbsent(sfDiscountedFee);  // LCOV_EXCL_LINE
 }
 
+bool
+isOnlyLiquidityProvider(
+    ReadView const& view,
+    AccountID const& ammAccount,
+    AccountID const& lpAccount)
+{
+    std::uint8_t nTrustLines = 0;
+    std::uint8_t limit = 10;
+    auto const root = keylet::ownerDir(ammAccount);
+    auto currentIndex = root;
+
+    // Iterate over AMM owner directory objects.
+    // AMM Liquidity Provider has at most three
+    // trustlines. One for LPToken, and at most
+    // two trustlines for IOU: one if XRP/IOU
+    // AMM and two if IOU/IOU AMM.
+    while (limit-- >= 1)
+    {
+        auto const ownerDir = view.read(currentIndex);
+        if (!ownerDir)
+            return false;
+        for (auto const& key : ownerDir->getFieldV256(sfIndexes))
+        {
+            auto const sle = view.read(keylet::child(key));
+            if (!sle)
+            {
+                assert(false);
+                return false;
+            }
+            // Only one AMM object
+            if (sle->getFieldU16(sfLedgerEntryType) == ltAMM)
+                continue;
+            if (sle->getFieldU16(sfLedgerEntryType) != ltRIPPLE_STATE)
+            {
+                assert(false);
+                return false;
+            }
+            auto const lowLimit = sle->getFieldAmount(sfLowLimit);
+            auto const highLimit = sle->getFieldAmount(sfHighLimit);
+            if ((lowLimit.getIssuer() != lpAccount &&
+                 highLimit.getIssuer() != lpAccount))
+                return false;
+            ++nTrustLines;
+        }
+        auto const uNodeNext = ownerDir->getFieldU64(sfIndexNext);
+        if (uNodeNext == 0)
+            return nTrustLines == 2 || nTrustLines == 3;
+        currentIndex = keylet::page(root, uNodeNext);
```

Comment:
This linear search is kinda terrifying.  I looked in a recent ledger and the largest number of owner directory pages owned by a single account was 284,820.  But there were quite a few accounts with over 10,000 owner directory pages.  This linear search signs us up for a potentially huge amount of computing.  Calling this function from a transactor looks really scary to me.

Is there anything we know about AMM Liquidity Providers that will limit the number of directory pages that they can own?  Is there any way that could be leveraged to limit this linear search to just a few directory nodes?

Reply:
There are two things in this function that prevent this possible huge number of iterations. One is that the limit is set to at most 10 iterations. Two is that if there is more than one Liquidity Provider (LP) then we'll break out of the loop a lot sooner because any LP can have at most three trustlines to AMM: one for LPToken and at most two for IOU. So we should break out after at most four iterations in the worst case scenario if there are more than one LP.
   
    if ((lowLimit.getIssuer() != lpAccount &&
                 highLimit.getIssuer() != lpAccount))
                return false;

Reply:
Sorry.  I missed what `limit` was up to when I first looked at the `while` loop.  But I still don't understand why 10 directory pages is a legit upper bound.

Owner directories can hold a lot of things other than trust lines.  The number of `Offer`s (both IOU and NFT), `Escrow`s, `PayChan`s, and `DepositPreauth`s they can hold has a very large upper bound.  Additionally, an account can own up to 250 tickets.

Now, maybe it's impossible for an AMM Liquidity Provider account to own any of these kinds of objects.  In which case 10 iterations is (probably?) plenty.  But I am not familiar enough with the AMM to know whether that's the case.  If it _is_ the case I think there should be a comment explaining why 10 iterations will always be sufficient.

There should probably also be some kind of error indication if the number of expected directory pages is ever surpassed.  It may not be possible for an AMM Liquidity Provider to ever exceed that number of pages today.  But people are excited about AMMs, which means rules will be changing in the future.  So either the function must throw or the return value needs to be more complicated than a `bool`.  Maybe an `Expected`?

Does that make any sense?

Reply:
We are iterating over the trustlines for AMM account. AMM account holds one AMM object and the rest are LP trustlines. Each LP has one trustline for LPToken. If an issuer created AMM then the issuer could have two more trustlines for IOU. Actually the function should return `true` if trustlines >= 1 && truslines <= 3, not 2 or 3.

It makes sense to return an error in some cases: there are 0 trustlines, non-issuer LP has more than 1 trustline, an issuer LP has more than 3 trustlines, there are more than 1 AMM object. `Expected` with `tecINTERNAL` works.

Reply:
I think it is sufficient to just look for LPToken trustlines and return an error if it is 0 or greater than 1.

Reply:
I just did a little more reading about the AMM implementation.  I'm sorry that a) I'm not really up on how the AMM works and b) that I did not look more closely at the loop before I commented.

Given the constraints as I (finally) understand them, we can have high confidence that the number of loops will be very bounded by adding one constraint to the loop.  If `nTrustLines > 3` return `false`.  I think if that constraint is added then (assuming there's only one AMM object) the number of directory pages examined cannot exceed 6.

Reply:
Yeah, I don't love this. I don't see a better solution though.